docs/extended-utxo/ValidatingTheNextDataScript.md

```diff
@@ -0,0 +1,236 @@
+
+It has this weird behavior: if `a` and `b` are definitionally the same, then `cast` is equal to `just`, otherwise it's equal to `const nothing`. We can't of course define such a function in Plutus Core, because it checks equality of types at runtime and we do not have a runtime representation of types, so we make `cast` a magic builtin.
+
+Having this function, we can make transaction output to be of the following type (I'm ignoring the value restriction problem for now):
+
+```
+list (hash × (all b. maybe b))
+```
+
+(where `hash` is an appropriate instantiation of `bytestring`). Here is an example (in pseudosyntax):
+
+```
+txOutput : list (hash × (all b. maybe b))
+txOutput =
+    [ ("aaa", /\b -> cast {integer 1}    {b} (1!1)  )
+    , ("bbb", /\b -> cast {unit}         {b} unitval)
+    , ("ccc", /\b -> cast {bool -> bool} {b} not    )
+    ]
+```
+
+And then we can do the option parsing thing. For example if you need to get the part of the transaction output that has hash "ccc" and type `bool -> bool`, you just look up "ccc", get
+
+```
+/\ b -> cast {bool -> bool} {b} not
+```
+
+and instantiate `b` with `bool -> bool`:
+
+```
+cast {bool -> bool} {bool -> bool} not
+```
+
+which computes to `just not` of type `maybe (bool -> bool)`.
+
+This way we get homogeneous lists that contain heterogeneous values inside, which we can extract by supplying expected types of those values.
+
+Note that `cast` is isomorphic to `dynamic`:
+
+- we can represent `dynamic` as `all b. maybe` b and implement it using `cast`
+- they seem equally powerful: we can implement `cast` as `fromDynamic . toDynamic`. But `cast` seems conceptually simpler
+
+so all the same problems apply.
+
+### Safe `cast`/`dynamic`
+
+There is however what I believe is a safe version of the `cast`/`dynamic` approach (they are isomorphic, so I'll be calling them the same approach).
+
+The reason why regular `cast`/`dynamic` breaks parametricity is because we're allowed to call `cast`/`dynamic` over an argument that has an open type (i.e. a type that contains free variables). This is what we do above in the `reverseNonIntegers` function. Once we forbid to use open types, control flow can't be affected by `cast`/`dynamic` during evaluation and no type-case is possible to implement. We still can write things like
+
+```
+case cast {integer} {bytestring} 1 of
+    nothing  -> ...
+	just one -> ...
+```
+
+but here everything we need in order to determine runtime control flow is known at compile time. This is like instance search in Haskell where runtime behavior of a program is affected by types, because they drive instance resolution, but the behavior is determined at compile time.
+
+(In fact, I believe we can even allow open types and still get parametricity by introducing a special monad to handle what is normally called quoting, but we do not really need this power to complicate the language)
+
+However we still have the problems of ensuring that types are closed and can be erased.
```

Comment:
Also, when we implement `dynamic` with `cast`, don't we need an open type as one of the arguments? i.e. `toDynamic 1 = forall a. cast {integer} {a} 1`.

I think having even one argument open is enough to get typecase.

Reply:
Hm, it seems I forgot to answer that question and now I don't know if there exists a satisfying answer.